sdks/java/io/xml/src/main/java/org/apache/beam/sdk/io/xml/XmlSource.java

```diff
@@ -281,7 +283,15 @@ private long getFirstOccurenceOfRecordElement(
               break outer;
             } else {
               // Matching was unsuccessful. Reset the buffer to include bytes read for the char.
-              ByteBuffer newbuf = ByteBuffer.allocate(BUF_SIZE);
+              int bytesToWrite = buf.remaining() + charBytes.length;
+              ByteBuffer newbuf;
+              if (bytesToWrite > BUF_SIZE) {
+                // Avoiding buffer overflow
```

Comment:
Probably expand comment to following.

"The number of bytes to push to the buffer might be larger than BUF_SIZE due to additional 'charBytes'."

Reply:
done